recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,401 @@
+
+### Key-value pairs
+Key-value pairs will be parsed and saved to a Terraform configuration file.
+
+### Secrets
+Secrets will be handled similarly to the approach described in document [Design document for Private Terraform Repository](https://github.com/radius-project/design-notes/blob/3644b754152edc97e641f537b20cf3d87a386c43/recipe/2024-01-support-private-terraform-repository.md) wherein Applications.Core/secretStores can point to an existing K8s secret.
+
+The system will call the ListSecrets() api in Applications.Core namespace, retrieve contents of the secret and build the Terraform provider configuration.
+
+### Environment variables
+In a significant number of providers, as per documentation, environment variables are used one of the methods of saving sensitive credential data along with insensitive data. We allow the users to set environment variables for provider configuration. For sensitive information, we recommend the users save these values as secrets and point to them inside the env block.
+
+```
+...
+ recipeConfig: {
+    terraform: {
+      ...
+     providers: [...]
+     env: {
+      key: value
+      envSecrets: [
+        {
+          name: 'SPECIAL_LEVEL_KEY'
+          valueFrom: {
+            source: secretStoreConfig.id
+            name: 'special-config'
+            key: 'special.level'
+          }
+        },
+        {
+          name: 'LOG_LEVEL'
+          valueFrom: {
+            source: secretStoreConfig.id
+            name: 'special-config'
+            key: 'special.log'
+          }
+        }
+      ]
+      envSecretsFrom: {                   // All Secrets from Secret Store
+        secrets: {                        // This exists:Private TF module
+          source: secretStoreOCICred.id
+        }
+      }
+    }
+```
+
+Environment variables apply to all providers configured in the environment. The system cannot set two separate values for the same environment variable for multiple provider configurations. In such cases, per provider documentation, there may be alternatives that users can avail (eg. For GCP, users can set  *credentials* field inside each instance of provider config as opposed to using env variable GOOGLE_APPLICATION_CREDENTIALS).
+
+
+The system will allow for ability to set up multiple configurations for the same provider using the keyword *alias*. Validation for configuration 'correctness' will be handled by Terraform with calls to terraform init and terraform apply.
+
+```
+...
+ recipeConfig: {
+    terraform: {
+      ...
+      providers: [
+      {
+        name: 'azurerm',
+        properties: {
+          subscriptionid: 1234,
+            secrets: {
+              source: secretStoreAz.id
+            }
+          }
+        },
+        {
+          name: 'azurerm',
+          properties: {
+            subscriptionid: 1234,
+            alias: 'az-paymentservice',
+            secrets: {
+              source: secretStoreAzPayment.id
+            }
```

Comment:
Is this map going to have any property other than **source**?

Reply:
updated schema per discussion